jlm/util/GraphWriter.hpp

```diff
@@ -0,0 +1,843 @@
+
+  /**
+   * @return true if this GraphElement has been finalized, otherwise false.
+   */
+  [[nodiscard]] bool
+  IsFinalized() const;
+
+  /**
+   * Prints the attributes of the graph element.
+   * @param out the stream output is written to.
+   * @param format the output format to use.
+   */
+  void
+  OutputAttributes(std::ostream & out, AttributeOutputFormat format);
+
+private:
+  void
+  SetProgramObjectUintptr(uintptr_t object);
+
+  // A human-readable piece of text that should be rendered with the element
+  std::string Label_;
+
+  // A number added to the end of the id stub to make it globally unique
+  int UniqueIdSuffix_;
+
+  // The object in the program this graph object corresponds to, or 0 if none
+  uintptr_t ProgramObject_;
+
+  // Arbitrary collection of other attributes. The value can be a string, a reference to a
+  // GraphElement, or a reference to a program object.
+  using AttributeValue = std::variant<std::string, GraphElement *, uintptr_t>;
+  std::unordered_map<std::string, AttributeValue> AttributeMap_;
+};
+
+/**
+ * Abstract class representing a part of a node that edges can be attached to
+ */
+class Port : public GraphElement
+{
+  friend Edge;
+
+protected:
+  Port();
+
+public:
+  ~Port() override = default;
+
+  virtual Node &
+  GetNode() = 0;
+
+  [[nodiscard]] Graph &
+  GetGraph() override;
+
+  /**
+   * @return true if a directed edge may have its head at this port, false otherwise
+   */
+  [[nodiscard]] virtual bool
+  CanBeEdgeHead() const;
+
+  /**
+   * @return true if a directed edge may have its tail at this port, false otherwise
+   */
+  [[nodiscard]] virtual bool
+  CanBeEdgeTail() const;
+
+  /**
+   * @return true if any edges are leaving this port, or any non-directed edges are present
+   */
+  [[nodiscard]] bool
+  HasOutgoingEdges() const;
+
+  /**
+   * @return true if any edges are pointing at this port, or any non-directed edges are present
+   */
+  [[nodiscard]] bool
+  HasIncomingEdges() const;
+
+  /**
+   * Helper function for setting the background color of the Port using the correct dot attributes
+   */
+  virtual void
+  SetFillColor(std::string color) = 0;
+
+  /**
+   * Outputs the fully qualified port name, such as node8:i6
+   * Only used by the Dot printer.
+   */
+  virtual void
+  OutputDotPortId(std::ostream & out) = 0;
+
+  /**
+   * Only used by the ASCII printer.
+   * Outputs the origin(s) of edges pointing to this port.
+   * Brackets are omitted when exactly one edge points to the port.
+   * Example outputs: "o4", "[]" and "[o2, o6]"
+   */
+  void
+  OutputIncomingEdgesASCII(std::ostream & out);
+
+private:
+  /**
+   * Called when an edge has been added to the graph, to inform the edge's ports
+   * @param edge the newly added edge
+   */
+  void
+  OnEdgeAdded(Edge & edge);
+
+  std::vector<Edge *> Connections_;
+};
+
+/**
+ * Abstract base class for all nodes in the GraphWriter.
+ * A Node is also a port, since edges can be attached to nodes directly.
```

Comment:
Thanks for the explanation. I was about to write a comment about the inheritance from port, then I read it.